sotodlib/site_pipeline/preprocess_obs.py

```diff
@@ -0,0 +1,231 @@
+        
+    if type(context) == str:
+        context = core.Context(context)
+    
+    # if context doesn't have the preprocess archive it in add it
+    # allows us to use same context before and after calculations
+    found=False
+    if context.get("metadata") is None:
+        context["metadata"] = []
+
+    for key in context.get("metadata"):
+        if key.get("name") == "preprocess":
+            found=True
+            break
+    if not found:
+        context["metadata"].append( 
+            {
+                "db" : configs["archive"]["index"],
+                "name" : "preprocess"
+            }
+        )
+    return configs, context
+
+def preprocess_obs(
+    obs_id, 
+    configs,  
+    overwrite=False, 
+    logger=None
+):
+    """Meant to be run as part of a batched script, this function calls the
+    preprocessing pipeline a specific Observation ID and saves the results in
+    the ManifestDb specified in the configs.   
+
+    Arguments
+    ----------
+    obs_id: string or ResultSet entry
+        obs_id or obs entry that is passed to context.get_obs
+    configs: string or dictionary
+        config file or loaded config directory
+    overwrite: bool
+        if True, overwrite existing entries in ManifestDb
+    logger: logging instance
+        the logger to print to
+    """
+
+    if logger is None: 
+        logger = sp_util.init_logger("preprocess")
+    
+    if type(configs) == str:
+        configs = yaml.safe_load(open(configs, "r"))
+
+    context = core.Context(configs["context_file"])
+ 
+    if os.path.exists(configs['archive']['index']):
+        logger.info(f"Mapping {configs['archive']['index']} for the "
+                    "archive index.")
+        db = core.metadata.ManifestDb(configs['archive']['index'])
+    else:
+        logger.info(f"Creating {configs['archive']['index']} for the "
+                     "archive index.")
+        scheme = core.metadata.ManifestScheme()
+        scheme.add_exact_match('obs:obs_id')
+        scheme.add_data_field('dataset')
+        db = core.metadata.ManifestDb(
+            configs['archive']['index'],
+            scheme=scheme
+        )
+
+    pipe = Pipeline(configs["process_pipe"], plot_dir=configs["plot_dir"], logger=logger)
+
+    logger.info(f"Beginning run for {obs_id}")
+
+    aman = context.get_obs(obs_id, no_signal=True)
+    proc_aman = pipe.run(aman)
+
+    policy = sp_util.ArchivePolicy.from_params(configs['archive']['policy'])
+    dest_file, dest_dataset = policy.get_dest(obs_id)
+    logger.info(f"Saving data to {dest_file}:{dest_dataset}")
+    proc_aman.save(dest_file, dest_dataset, overwrite=overwrite)
+
+    # Update the index.
+    db_data = {'obs:obs_id': obs_id,
+                'dataset': dest_dataset}
+    
+    logger.info(f"Saving to database under {db_data}")
+    if len(db.inspect(db_data)) == 0:
+        db.add_entry(db_data, dest_file)
+
+def load_preprocess_det_select(obs_id, configs, context=None):
```

Comment:
Is detector selection something that's necessary for observation level analysis?